src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs

```diff
@@ -391,22 +391,61 @@ public void CopyTo(T[] array, int arrayIndex)
             Array.Copy(_items, 0, array, arrayIndex, _size);
         }
 
-        // Ensures that the capacity of this list is at least the given minimum
-        // value. If the current capacity of the list is less than min, the
-        // capacity is increased to twice the current capacity or to min,
-        // whichever is larger.
-        //
-        private void EnsureCapacity(int min)
+        /// <summary>
+        /// Ensures that the capacity of this list is at least the specified <paramref name="capacity"/>.
+        /// If the current capacity of the list is less than specified <paramref name="capacity"/>,
+        /// the capacity is increased by continuously twice current capacity until it is at least the specified <paramref name="capacity"/>.
+        /// </summary>
+        /// <param name="capacity">The minimum capacity to ensure</param>
+        public int EnsureCapacity(int capacity)
         {
-            if (_items.Length < min)
+            if (capacity < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.capacity, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            if (_items.Length < capacity)
             {
-                int newCapacity = _items.Length == 0 ? DefaultCapacity : _items.Length * 2;
-                // Allow the list to grow to maximum possible capacity (~2G elements) before encountering overflow.
-                // Note that this check works even when _items.Length overflowed thanks to the (uint) cast
-                if ((uint)newCapacity > Array.MaxArrayLength) newCapacity = Array.MaxArrayLength;
+                IncreaseCapacityTo(capacity);
+                _version++;
+            }
+
+            return _items.Length;
+        }
+
+        private void EnsureCapacityWithoutUpdatingVersion(int capacity)
+        {
+            if (_items.Length < capacity)
+            {
+                IncreaseCapacityTo(capacity);
+            }
+        }
+
+        /// <summary>
+        /// Increase the capacity of this list to at least the specified <paramref name="min"/> by continuously twice current capacity.
+        /// </summary>
+        /// <param name="min">The minimum capacity to ensure</param>
+        private void IncreaseCapacityTo(int min)
+        {
+            int newCapacity = _items.Length == 0 ? DefaultCapacity : _items.Length * 2;
+            while (newCapacity < min)
+            {
+                newCapacity *= 2;
+            }
+            // Allow the list to grow to maximum possible capacity (~2G elements) before encountering overflow.
+            // Note that this check works even when _items.Length overflowed thanks to the (uint) cast
+            if ((uint)newCapacity > Array.MaxArrayLength)
+            {
+                newCapacity = Array.MaxArrayLength;
                 if (newCapacity < min) newCapacity = min;
-                Capacity = newCapacity;
             }
+
```

Comment:
This should use the existing `Capacity` setter instead.

Reply:
It is based on [comment](https://github.com/dotnet/runtime/pull/47149#discussion_r560239273)

Reply:
Oops. See my previous comment, would rather take one extra check over duplicating this logic. My original feedback might have led you down this path, apologies for the confusion.

Reply:
No problem, just confirm: this will also use existing Capacity setter even though it will introduce more than one extra check.

Reply:
Yes, it is preferable to duplicating the logic.

Reply:
Fixed.